src/contradictory_claims/models/sbert_models.py

```diff
@@ -0,0 +1,382 @@
+    train_dataloader_embed = DataLoader(train_data, shuffle=True, batch_size=batch_size)
+    train_loss_embed = losses.SoftmaxLoss(
+        model=model.embedding_model,
+        sentence_embedding_dimension=model.embedding_model.get_sentence_embedding_dimension(),
+        num_labels=train_num_labels)
+
+    val_nli_reader = NLIDataReader(df_val)
+    dev_data = SentencesDataset(val_nli_reader.get_examples(), model=model.embedding_model)
+    dev_data.label_type = torch.long
+    evaluator = EmbeddingSimilarityEvaluator(sentences1=df_val["sentence1"].values,
+                                             sentences2=df_val["sentence2"].values,
+                                             scores=df_val["label"].values / 2.,
+                                             batch_size=batch_size)
+    warmup_steps = math.ceil(len(train_dataloader_embed) * epochs / batch_size * 0.1)  # 10% of train data for warm-up
+
+    # now to train the final layer
+    train_dataset = ClassifierDataset(df_train, tokenizer=tokenizer)
+    val_dataset = ClassifierDataset(df_val, tokenizer=tokenizer)
+    if enable_class_weights==False:
+        class_weights=None
+    else:
+        class_weights = train_dataset.class_weights()
+
+    train_dataloader = DataLoader(dataset=train_dataset,
+                                  batch_size=batch_size,
+                                  collate_fn=collate_fn,
+                                  shuffle=True)
+    val_dataloader = DataLoader(dataset=val_dataset,
+                                batch_size=1,
+                                collate_fn=collate_fn)
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+    criterion = nn.CrossEntropyLoss(weight=class_weights.to(device))
+    optimizer = optim.Adam(model.parameters(), lr=learning_rate)
+    model.to(device)
+    accuracy_stats = {"train": [],
+                      "val": [],
+                      }
+    loss_stats = {"train": [],
+                  "val": [],
+                  }
+
+    print("------TRAINING STARTS----------")  # noqa: T001
+    ## train embedding layer
+    unfreeze_layer(model.embedding_model)
+    model.embedding_model.fit(train_objectives=[(train_dataloader_embed, train_loss_embed)],
+                              evaluator=evaluator,
+                              epochs=1,
+                              evaluation_steps=1000,
+                              warmup_steps=warmup_steps,
+                              )  # train the Transformer layer
+    freeze_layer(model.embedding_model)
+
+
+    for e in range(epochs):
+        train_epoch_loss = 0
+        train_epoch_acc = 0
+        model.train()
+
+        for sentence1, sentence2, label in tqdm(train_dataloader):
+            label = label.to(device)
+            optimizer.zero_grad()
+            y_train_pred = model(sentence1, sentence2)
+
+            train_loss = criterion(y_train_pred, label)
```

Comment:
If I understand right, "y_train_pred" is the embedding output and needs to go through a softmax layer before this step...
Also, not sure if this whole section is really needed (previous comment).

Reply:
the criterion already has log_softmax inbuilt, so it's doing what we intend it to do

Reply:
Hmm...seeing from line 160 of this file, "criterion" is the crossentropyloss, which means it expects 2 probability distributions as inputs. I don't see how it has a log_softmax inbuilt? Could you clarify?

Reply:
check this https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html